File: wallet-web/app/src/api/cryptoApi.ts
// Client for wallet-backend's crypto component (PLAN.md §18/19):
// OneLiquidity-backed external-crypto deposit/withdrawal. Withdrawal is a
// real Safe transaction (the treasury debit) - build -> personal_sign
// digest -> confirm, same pattern as paymentsApi.ts, per wallet-backend
// PLAN.md §18's fix.
import { apiRequest } from './httpClient';

export interface DepositAddress {
  address: string;
  network: string;
}

export interface CryptoDeposit {
  currency: string;
  amount: string;
  txHash: string;
  createdAt: string;
}

export interface WithdrawalNetwork {
  network: string;
  withdrawMin: string;
  withdrawMax: string;
  withdrawFee: string;
}

export interface WithdrawalNetworksResult {
  networks: WithdrawalNetwork[];
  treasuryAddress: string;
}

export interface WithdrawalProposal {
  actionId: number;
  digestToSign: string;
}

export interface WithdrawalRequest {
  id: string;
  currency: string;
  network: string;
  toAddress: string;
  amountSubmitted: number;
  amountToWithdraw: number;
  status: string;
  treasuryTxHash: string;
  createdAt: string;
}

export function getDepositAddresses(walletAddress: string, currency: string): Promise<DepositAddress[]> {
  return apiRequest<DepositAddress[]>(`/v1/crypto/deposit-address/${encodeURIComponent(currency)}`, { walletAddress });
}

export function getDepositHistory(walletAddress: string): Promise<CryptoDeposit[]> {
  return apiRequest<CryptoDeposit[]>('/v1/crypto/deposit-history', { walletAddress });
}

export function getWithdrawalNetworks(walletAddress: string, currency: string): Promise<WithdrawalNetworksResult> {
  return apiRequest<WithdrawalNetworksResult>(`/v1/crypto/withdrawal-networks/${encodeURIComponent(currency)}`, {
    walletAddress,
  });
}

export function buildWithdrawal(
  walletAddress: string,
  currency: string,
  network: string,
  amount: number,
): Promise<WithdrawalProposal> {
  return apiRequest<WithdrawalProposal>('/v1/crypto/withdrawals/build', {
    method: 'POST',
    walletAddress,
    body: { currency, network, amount },
  });
}

export function confirmWithdrawal(
  walletAddress: string,
  currency: string,
  network: string,
  toAddress: string,
  amount: number,
  actionId: number,
  signature: string,
): Promise<WithdrawalRequest> {
  return apiRequest<WithdrawalRequest>('/v1/crypto/withdrawals/confirm', {
    method: 'POST',
    walletAddress,
    body: { currency, network, toAddress, amount, actionId, signature },
  });
}

export function getWithdrawalHistory(walletAddress: string): Promise<WithdrawalRequest[]> {
  return apiRequest<WithdrawalRequest[]>('/v1/crypto/withdrawal-history', { walletAddress });
}

File: wallet-mobile/app/lib/api/crypto_api.dart
// Mirrors wallet-web's api/cryptoApi.ts - client for wallet-backend's
// crypto component (PLAN.md §18/19): OneLiquidity-backed external-crypto
// deposit/withdrawal. Withdrawal is a real Safe transaction (the treasury
// debit) - build -> personal_sign digest -> confirm, same pattern as
// payments_api.dart, per wallet-backend PLAN.md §18's fix.
import 'http_client.dart';

class DepositAddress {
  DepositAddress({required this.address, required this.network});
  final String address;
  final String network;

  factory DepositAddress.fromJson(Map<String, dynamic> json) =>
      DepositAddress(address: json['address'] as String, network: json['network'] as String);
}

class WithdrawalNetwork {
  WithdrawalNetwork({required this.network, required this.withdrawMin, required this.withdrawMax, required this.withdrawFee});
  final String network;
  final String withdrawMin;
  final String withdrawMax;
  final String withdrawFee;

  factory WithdrawalNetwork.fromJson(Map<String, dynamic> json) => WithdrawalNetwork(
        network: json['network'] as String,
        withdrawMin: json['withdrawMin'] as String,
        withdrawMax: json['withdrawMax'] as String,
        withdrawFee: json['withdrawFee'] as String,
      );
}

class WithdrawalNetworksResult {
  WithdrawalNetworksResult({required this.networks, required this.treasuryAddress});
  final List<WithdrawalNetwork> networks;
  final String treasuryAddress;

  factory WithdrawalNetworksResult.fromJson(Map<String, dynamic> json) => WithdrawalNetworksResult(
        networks: (json['networks'] as List).map((e) => WithdrawalNetwork.fromJson(e as Map<String, dynamic>)).toList(),
        treasuryAddress: json['treasuryAddress'] as String,
      );
}

class WithdrawalProposal {
  WithdrawalProposal({required this.actionId, required this.digestToSign});
  final int actionId;
  final String digestToSign;

  factory WithdrawalProposal.fromJson(Map<String, dynamic> json) =>
      WithdrawalProposal(actionId: json['actionId'] as int, digestToSign: json['digestToSign'] as String);
}

class WithdrawalRequest {
  WithdrawalRequest({
    required this.id,
    required this.currency,
    required this.network,
    required this.toAddress,
    required this.amountSubmitted,
    required this.amountToWithdraw,
    required this.status,
    required this.treasuryTxHash,
    required this.createdAt,
  });

  final String id;
  final String currency;
  final String network;
  final String toAddress;
  final num amountSubmitted;
  final num amountToWithdraw;
  final String status;
  final String treasuryTxHash;
  final String createdAt;

  factory WithdrawalRequest.fromJson(Map<String, dynamic> json) => WithdrawalRequest(
        id: json['id'] as String,
        currency: json['currency'] as String,
        network: json['network'] as String,
        toAddress: json['toAddress'] as String,
        amountSubmitted: json['amountSubmitted'] as num,
        amountToWithdraw: json['amountToWithdraw'] as num,
        status: json['status'] as String,
        treasuryTxHash: json['treasuryTxHash'] as String,
        createdAt: json['createdAt'] as String,
      );
}

class CryptoApi {
  CryptoApi(this._client);
  final ApiClient _client;

  Future<List<DepositAddress>> getDepositAddresses(String walletAddress, String currency) {
    return _client.request(
      '/v1/crypto/deposit-address/${Uri.encodeComponent(currency)}',
      walletAddress: walletAddress,
      decode: (json) => (json as List).map((e) => DepositAddress.fromJson(e as Map<String, dynamic>)).toList(),
    );
  }

  Future<WithdrawalNetworksResult> getWithdrawalNetworks(String walletAddress, String currency) {
    return _client.request(
      '/v1/crypto/withdrawal-networks/${Uri.encodeComponent(currency)}',
      walletAddress: walletAddress,
      decode: (json) => WithdrawalNetworksResult.fromJson(json as Map<String, dynamic>),
    );
  }

  Future<WithdrawalProposal> buildWithdrawal(String walletAddress, String currency, String network, num amount) {
    return _client.request(
      '/v1/crypto/withdrawals/build',
      method: 'POST',
      walletAddress: walletAddress,
      body: {'currency': currency, 'network': network, 'amount': amount},
      decode: (json) => WithdrawalProposal.fromJson(json as Map<String, dynamic>),
    );
  }

  Future<WithdrawalRequest> confirmWithdrawal(
    String walletAddress,
    String currency,
    String network,
    String toAddress,
    num amount,
    int actionId,
    String signature,
  ) {
    return _client.request(
      '/v1/crypto/withdrawals/confirm',
      method: 'POST',
      walletAddress: walletAddress,
      body: {
        'currency': currency,
        'network': network,
        'toAddress': toAddress,
        'amount': amount,
        'actionId': actionId,
        'signature': signature,
      },
      decode: (json) => WithdrawalRequest.fromJson(json as Map<String, dynamic>),
    );
  }
}
